contracts/lib/OnchainNonMerklizedIdentityBase.sol

```diff
@@ -0,0 +1,245 @@
+            mtp: mtp
+        });
+        IssuanceProof[] memory proofs = new IssuanceProof[](1);
+        proofs[0] = issuanceProof;
+
+        uint256 chainId;
+        assembly {
+            chainId := chainid()
+        }
+        return
+            CredentialData({
+                id: string(
+                    abi.encodePacked(
+                        "urn:onchain:",
+                        Strings.toString(chainId),
+                        ":",
+                        Strings.toHexString(uint160(address(this))),
+                        ":",
+                        Strings.toString(credentialMetadata.id)
+                    )
+                ),
+                context: fillContext(credentialInformation.jsonLDContextUrls),
+                _type: fillType(credentialInformation._type),
+                expirationDate: credentialMetadata.expirationDate,
+                issuanceDate: credentialMetadata.issuanceDate,
+                issuer: identity.id,
+                credentialSubject: fillCredentialSubject(
+                    credentialSubject,
+                    credentialInformation._type
+                ),
+                credentialStatus: Status({
+                    id: "/credentialStatus",
+                    _type: "Iden3OnchainSparseMerkleTreeProof2023",
+                    revocationNonce: credentialMetadata.revocationNonce
+                }),
+                credentialSchema: Schema({
+                    id: credentialInformation.jsonSchemaUrl,
+                    _type: "JsonSchema2023"
+                }),
+                proof: proofs
+            });
+    }
+
+    function fillContext(string[] memory context) private pure returns (string[] memory) {
+        string[] memory newContext = new string[](context.length + 2);
+        for (uint256 i = 0; i < context.length; i++) {
+            newContext[i] = context[i];
+        }
+        newContext[context.length] = "https://www.w3.org/2018/credentials/v1";
+        newContext[context.length + 1] = "https://schema.iden3.io/core/jsonld/iden3proofs.jsonld";
+        return newContext;
+    }
+
+    function fillType(string memory _type) private pure returns (string[] memory) {
+        string[] memory newType = new string[](2);
+        newType[0] = "VerifiableCredential";
+        newType[1] = _type;
+        return newType;
+    }
+
+    function fillCredentialSubject(
+        SubjectField[] memory fields,
+        string memory _type
+    ) private pure returns (SubjectField[] memory) {
+        SubjectField[] memory credentialSubjectWithDefaultFields = new SubjectField[](
+            fields.length + 1
+        );
+        for (uint256 i = 0; i < fields.length; i++) {
+            credentialSubjectWithDefaultFields[i] = fields[i];
+        }
+        credentialSubjectWithDefaultFields[fields.length] = SubjectField({
+            key: "type",
+            value: 0,
+            rawValue: bytes(_type)
+        });
```

Comment:
I think you also need to add "id" attribute if it's not empty.

Reply:
id field is not a required field for credential subjects. If a client doesn't pass the ID, the client wants to issue a self-credential.

Reply:
Fixed